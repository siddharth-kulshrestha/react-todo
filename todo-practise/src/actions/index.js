// Types of action
export const Types = {
    CREATE_ITEM: "CREATE_ITEM",
    UPDATE_RATING: "UPDATE_RATING",
    ADD_ITEMS: "ADD_ITEMS",
}

// createItem action
export const createItem = item => ({
    type: Types.CREATE_ITEM,
    payload: item
});

// addMultipleItems action
export const addMultipleItems = items => ({
    type: Types.ADD_ITEMS,
    payload: items,
})

// addMultipleItems action
export const updateRating = (id, newRating) => ({
    type: Types.UPDATE_RATING,
    payload: {itemID: id, rating: newRating},
})
