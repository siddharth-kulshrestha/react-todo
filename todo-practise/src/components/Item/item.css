.item-container {
    border-style: solid;
    border-radius: 5%;
    border-color: black;
    justify-content: space-around;
    align-items: center;
    flex-direction: row;
    display: flex;
    flex: 1;
    
}

.item-content {
    display: flex;
    flex-direction: row;
    flex: 2;
    margin: 10px;
}

.item-rate-handler {
    display: flex;
    flex-direction: row;
    flex: 1;
}

.rating-container {
    display: flex;
    flex-direction:row;
    justify-content: space-between;
}