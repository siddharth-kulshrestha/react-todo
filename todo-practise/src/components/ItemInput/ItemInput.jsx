import React, { PureComponent } from 'react';
import './iteminput.css';

export default class ItemInput extends PureComponent {
    state = {
        content: "",
        rating: 0,
    }

    handleContentChange(event) {
        this.setState({content: event.target.value});
    }

    handleRatingChange(event) {
        this.setState({rating: event.target.value});
    }

    handleOnSubmit(event) {
        event.preventDefault();
        this.props.onInput({content: this.state.content, rating: this.state.rating});
    }

    render() {
        return (
            <form className="form-container" onSubmit={this.handleOnSubmit.bind(this)}>
                <input 
                type="text"
                placeholder="Item Content"
                value={this.state.content}
                onChange={this.handleContentChange.bind(this)}
                />
                <input 
                type="number"
                placeholder="Rating"
                value={this.state.rating}
                onChange={this.handleRatingChange.bind(this)}
                />
                <input
                type="submit"
                value="Add"
                />
            </form>
        );
    }
}
