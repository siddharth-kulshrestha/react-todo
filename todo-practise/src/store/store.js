import { createStore } from 'redux';
import todoReducer from '../reducers';

export default function store(initialState) {
    const store = createStore(todoReducer, initialState);
    return store; 
};
