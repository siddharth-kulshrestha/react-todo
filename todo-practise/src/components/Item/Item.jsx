import React, { PureComponent } from 'react';
import './item.css';
import { Button } from '../Button/Button';

class RateItem extends PureComponent {

    state = {
        rating: 0,
    };

    componentDidMount() {
        if (this.props.rating !== undefined) {
            this.setState({rating: this.props.rating});
        }
    }

    onDecreaseRating() {
        this.setState((prevState) => ({rating: prevState.rating - 1}), this.onRatingChange);
    }

    onIncreaseRating() {
        this.setState((prevState) => ({rating: prevState.rating + 1}), this.onRatingChange);
    }

    onRatingChange() {
        if (this.props.onRatingChange) {
            this.props.onRatingChange(this.state.rating);
        }
    }

    render() {
        return (
            <div className="rating-container">
                <Button onClick={this.onDecreaseRating.bind(this)} text="-"/> {this.state.rating} <Button onClick={this.onIncreaseRating.bind(this)} text="+"/>
            </div>
        );
    }
}

export default class Item extends PureComponent {
    render() {
        console.log("Item-ID: "+this.props.data.id);
        console.log(this.props);
        return (
            <div className="item-container">
                <div className="item-content"><h3>{this.props.data.item.content}</h3></div> 
                <div className="item-rate-handler"><RateItem onRatingChange={(val) => {this.props.onRatingChange(this.props.data.item.id, val)}} rating={this.props.data.item.rating}/></div>
            </div>
        );
    }
}

