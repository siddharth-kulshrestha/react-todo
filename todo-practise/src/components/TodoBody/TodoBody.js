import React, { PureComponent } from 'react';
import ItemList from '../ItemList/ItemList';
import ItemInput from '../ItemInput/ItemInput';
import { connect } from 'react-redux'
import { createItem, addMultipleItems, updateRating } from '../../actions'

function getData() {
    console.log("Function to get data called!!");
    return [
        {
            content: "XYZ",
            rating: 6,
        },
        {
            content: "ABC",
            rating: 7,
        },
        {
            content: "PWQ",
            rating: 8,
        },
        {
            content: "MBQ",
            rating: 9,
        },
        {
            content: "CQLAZ",
            rating: 10,
        }
    ];
};

export class TodoBody extends PureComponent {
    state = {
        data: [],
    }

    componentDidMount() {
        const externalData = getData();
        // this.setState({data: externalData.slice()});
        this.props.dispatch(addMultipleItems(externalData))
    }

    onInputItem(item) {
        console.log(item);
        this.props.dispatch(createItem(item));
    }

    onItemRatingChange(id, newRating) {
        this.props.dispatch(updateRating(id, newRating))
    }

    render() {
        console.log('Props:');
        console.log(this.props);
        return (
            <div className="todo-body-container">
                <ItemInput onInput={this.onInputItem.bind(this)} />
                {this.props.data ? <ItemList data={this.props.data} onRatingChange={this.onItemRatingChange.bind(this)} /> : null}
            </div>
        );
    }
}

const mapStateToProps = function (store) {
    console.log('Servicedetails mapStatetoprops =',store );
    return {
        data: store.items,
    };
};

export default connect(mapStateToProps)(TodoBody)
