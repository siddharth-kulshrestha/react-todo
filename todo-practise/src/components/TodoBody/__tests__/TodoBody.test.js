import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import Adapter from 'enzyme-adapter-react-16';
import TodoBody from '../TodoBody';
import store from '../../../store/store';
import { createItem } from '../../../actions'
import ItemInput from '../../ItemInput/ItemInput';
import Item from '../../Item/Item';
import { SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS } from 'constants';


configure({adapter: new Adapter()});

const mockItem = {
    id: 0,
    item: {
        content: "test content",
        rating: 6,
    }
};

// TodoBody test suite testing actions with redux-mock-store
describe('TodoBody Test suite ', () => {
    const initialState = {
        items: []
    };
    const mockStore = configureStore();
    let store, component;

    beforeEach(() => {
        store = mockStore(initialState);
        component = mount( // componentDidMount() will not get trigger if we use shallow
        <Provider store={store}>
            <TodoBody />
        </Provider>
        );
    });

    it('check whether add multiple action call ', () => {
        const actions = store.getActions();
        expect(actions[0].type).toBe('ADD_ITEMS');
        expect(component).toMatchSnapshot();
    });

    it(' check whether add item action is getting called when an item is added ', () => {
        const itemInput = component.find(ItemInput);
        itemInput.setState({ content: "Test content", rating: 10 });
        
        const form = itemInput.find('form')
        form.simulate('submit', {
            preventDefault: () => {
            }
        });
        const actions = store.getActions();
        expect(actions[0].type).toBe('ADD_ITEMS');
        expect(actions[1].type).toBe('CREATE_ITEM');
        expect(actions[1].payload).toMatchObject({ content: "Test content", rating: 10 });
    });
});

// TodoBody test suite testing actions with actual store
describe('TodoBody Test suite ', () => {
    const initialState = {
        items: []
    };
    // const actualStore = store(initialState);
    let reduxStore, component;

    beforeEach(() => {
        // store = actualStore(initialState);
        reduxStore = store(initialState);
        component = mount( // componentDidMount() will not get trigger if we use shallow
        <Provider store={reduxStore}>
            <TodoBody />
        </Provider>
        );
    });

    it('test CREATE_ITEM action ', () => {
        reduxStore.dispatch(createItem({
            content: "test content",
            rating: 6,
        }));
        const storeState = reduxStore.getState();
        const lstIdx = storeState.items.length - 1;
        expect(storeState.items[lstIdx].item.content).toBe("test content");
        expect(storeState.items[lstIdx].item.rating).toBe(6);
    });

    it(' check whether add item action is getting called when an item is added ', () => {
        const itemInput = component.find(ItemInput);
        itemInput.setState({ content: "Test content 2 <unique>", rating: 10 });
        
        const form = itemInput.find('form')
        form.simulate('submit', {
            preventDefault: () => {
            }
        });

        // Verifying whether added item is added to store or not
        const storeState = reduxStore.getState();
        const lstIdx = storeState.items.length - 1;
        expect(storeState.items[lstIdx].item.content).toBe("Test content 2 <unique>");
        expect(storeState.items[lstIdx].item.rating).toBe(10);

        // Verifying added item in item-list
        const item = component.findWhere((comp) => comp.type() == Item &&
            comp.props().data && comp.props().data.item.content === 'Test content 2 <unique>'
            && comp.props().data.item.rating === 10);
        expect(item.length).toBe(1);
    });
});
