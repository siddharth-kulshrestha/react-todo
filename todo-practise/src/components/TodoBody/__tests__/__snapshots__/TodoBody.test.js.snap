// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoBody Test suite  check whether add multiple action call  1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(TodoBody)>
    <TodoBody
      data={Array []}
      dispatch={[Function]}
    >
      <div
        className="todo-body-container"
      >
        <ItemInput
          onInput={[Function]}
        >
          <form
            className="form-container"
            onSubmit={[Function]}
          >
            <input
              onChange={[Function]}
              placeholder="Item Content"
              type="text"
              value=""
            />
            <input
              onChange={[Function]}
              placeholder="Rating"
              type="number"
              value={0}
            />
            <input
              type="submit"
              value="Add"
            />
          </form>
        </ItemInput>
        <ItemList
          data={Array []}
          onRatingChange={[Function]}
        >
          <div
            className="item-list-container"
          />
        </ItemList>
      </div>
    </TodoBody>
  </Connect(TodoBody)>
</Provider>
`;
